### PR TITLE
Hive location widget

### PR DESCRIPTION
**Pull request by adzialocha**
This PR implements all models and UI widgets to set a hive location for a sighting.

#### Preview

Screenshot_20240520_150422
Screenshot_20240520_150411
Screenshot_20240520_153241

#### Schema Migration

```
bee_attributes_location_tree_0020a960a5e60c7daf66f9b9056de6e7904247360017dd09c223800e65223c0adafe

Name: bee_attributes_location_tree
Description: "Bee sighting location: tree"
╭───┬──────────────┬────────────────────────╮
│ # ┆ Field Name   ┆ Field Type             │
╞═══╪══════════════╪════════════════════════╡
│ 1 ┆ tree_species ┆ str                    │ <-- added
├╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 2 ┆ diameter     ┆ float                  │ <-- added
├╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 3 ┆ height       ┆ float                  │ <-- added
├╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 4 ┆ sighting     ┆ relation(bee_sighting) │
╰───┴──────────────┴────────────────────────╯
```

### CODE REVIEW

File: packages/app/lib/models/location.dart [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+  Building,
+  Ground,
+  Tree,
+}
+
+/*
+ * Generic location methods across all location types
+ */
+
+class Location {
```
Comment (sandreae) [inline packages/app/lib/models/location.dart:25]: Nice encapsulation of the different location types :+1:

File: packages/app/lib/models/location.dart [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+          $locationGroundFields
+        }
+      }
+      $TREE_RESULTS_KEY: all_$locationTreeSchemaId($parameters) {
+        documents {
+          $locationTreeFields
+        }
+      }
+    }
+  ''';
```
Comment (sandreae) [inline packages/app/lib/models/location.dart:192]: Again, very neat approach :+1:

File: packages/app/lib/ui/widgets/hive_location_field.dart [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+    super.initState();
+    location = widget.initialValue;
+  }
+
+  void _handleUpdate(LocationType? type, String? treeSpecies,
+      double? treeHeight, double? treeDiameter) async {
+    if (location != null && type == null) {
+      // Delete attached sighting if any existed
+      await location!.delete();
+      location = null;
```
Comment (sandreae) [inline packages/app/lib/ui/widgets/hive_location_field.dart:120]: Is there a mistake in the comment here?

Comment (sandreae) [reply packages/app/lib/ui/widgets/hive_location_field.dart:120]: "sighting" should be "location"

**Review by sandreae — APPROVED** [review]
Really good job combining the different location types (with their unique fields) into a single flow, and it's still easy to follow :+1: 

This all looks good to me except one possible mistake in a comment I pointed out.

**Comment by sandreae** [comment]
Also love the UI, veeeeery simple.

**Review by achou11 — COMMENTED** [review]
curious what the `HiveLocationTypeSelector` looks like when overflow occurs? I see the implementation uses `Wrap` which I guess means it'll just move onto another line. wondering if it would make sense to do a horizontal scroll on overflow, and change the minimum width of each button to make them more square by default.

**Pull request merged**